File: tokenizer.py
from collections import defaultdict
import time

CODE_POINT_START = 256  # Start value for custom code point assignment

def timed(func):
    def wrapper(*args, **kwargs):
        start = time.perf_counter()
        result = func(*args, **kwargs)
        elapsed = time.perf_counter() - start
        print(f"{func.__name__} took {elapsed:.4f} seconds")
        return result
    return wrapper


class Tokenizer():
    def __init__(self):
        self.data = ""
        self.vocab = {} 
 
    @timed
    def train(self, training_data):
        # Build initial vocabulary
        self.data = training_data
        self.vocab = {
            b: b
            for b in set(self.data.encode("utf-8"))
        } 

        # Convert chars to bytes
        byte_list = list(self.data.encode("utf-8"))

        unused_code_pt = CODE_POINT_START 

        # Byte pair loop
        while True:
            # Gather most frequent pairs
            pairs = defaultdict(int) 
            for b1, b2 in zip(byte_list, byte_list[1:]):
                pairs[(b1, b2)] += 1

            if not pairs:
                break

            max_freq = max(pairs.values()) 
            max_pair = max(pairs, key=pairs.get)

            if max_freq == 1:
                break

            new_byte_list = []
            for b1, b2 in zip(byte_list, byte_list[1:]):
                if (b1, b2) == max_pair:
                    new_byte_list.append(unused_code_pt)
                else:
                    new_byte_list.append(b1)

            byte_list = new_byte_list
            self.vocab[unused_code_pt] = max_pair 
            unused_code_pt += 1

    @timed
    def encode(self, s: str) -> list[int]: 
        """ Convert string to tokens """
        byte_list = list(s.encode("utf-8"))
        pair_found = True

        while pair_found:
            pair_found = False
            for i, s_pair in enumerate(zip(byte_list, byte_list[1:])):
                for code_pt, vocab_pair in self.vocab.items():
                    if s_pair == vocab_pair:
                        byte_list[i] = -1
                        byte_list[i + 1] = code_pt
                        pair_found = True
                    
            byte_list = [b for b in byte_list if b != -1]

        return byte_list
    
    @timed
    def decode(self, tokens) -> str:
        """ Convert tokens to string """
        pair_inserted = True
        final_code_points = []

        while pair_inserted:
            final_code_points = []
            pair_inserted = False

            for pt in tokens:
                if pt <= CODE_POINT_START - 1 or pt not in self.vocab:
                    # Only values above 255 are assigned pairs
                    final_code_points.append(pt)
                    continue

                b1, b2 = self.vocab[pt]
                final_code_points.extend([b1, b2])
               
                # Reset list - keep the loop going
                tokens = final_code_points
                pair_inserted = True

        # Convert code points to byte ints
        return self.byte_list_to_str(final_code_points)

    def byte_list_to_str(self, byte_list: list[int]) -> str:
        output_str = ""

        # Loop needed to handle multi-byte chars
        i = 0
        while i < len(byte_list):
            # Convert byte to binary string
            b_bin = f"{byte_list[i]:08b}"
            if b_bin[0] == "0":
                # Single byte sequence
                output_str += chr(byte_list[i])
                i += 1
                continue

            bytez = []
            offset = 0
            if all(b == '1' for b in b_bin[:4]):
                offset = 4  # 4 byte sequence
            elif all(b == '1' for b in b_bin[:3]):
                offset = 3  # 3 byte sequence
            else:
                offset = 2  # 2 byte sequence

            bytez = byte_list[i:i+offset]
            i += offset
            
            c = bytes(bytez).decode("utf-8")
            output_str += c

        return output_str 


print("\n" + "=" * 70)
print("TOKENIZER MAIN EXECUTION")
print("=" * 70)

data = ""
with open("training_text.txt", "r", encoding="utf-8") as f:
    data = f.read()

print(f"INPUT: Loaded from training_text.txt")
print(f"       Length: {len(data)} characters, {len(data.encode('utf-8'))} bytes")

print("\n--- TRAINING ---")
t = Tokenizer()
t.train(training_data=data)
print(f"Vocabulary size: {len(t.vocab)} tokens")
print(f"Learned merges: {len([k for k in t.vocab.keys() if k >= 256])}")

print("\n--- ENCODING ---")
encoded_data = t.encode(data)
print(f"Token count: {len(encoded_data)}")
compression_ratio = (1 - len(encoded_data) / len(data.encode('utf-8'))) * 100
print(f"Compression: {compression_ratio:.1f}% ({len(data.encode('utf-8'))} bytes → {len(encoded_data)} tokens)")

print("\n--- DECODING ---")
decoded_data = t.decode(encoded_data)
print(f"Output length: {len(decoded_data)} characters")

print("\n--- RESULT ---")
if data == decoded_data:
    print("✓ SUCCESS: Input matches output")
else:
    print("✗ FAILED: Input does not match output")
    print(f"  Expected length: {len(data)}")
    print(f"  Got length: {len(decoded_data)}")

print("=" * 70 + "\n")

File: tokenizer.rs
use std::collections::HashMap;
use std::fs;
use std::time::Instant;

const CODE_POINT_START: i32 = 256;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct Pair {
    b1: i32,
    b2: i32,
}

#[derive(Debug, Clone)]
enum VocabEntry {
    Byte(i32),
    Pair(Pair),
}

struct Tokenizer {
    vocab: HashMap<i32, VocabEntry>,
}

impl Tokenizer {
    fn new() -> Self {
        Tokenizer {
            vocab: HashMap::new(),
        }
    }

    fn train(&mut self, training_data: &str) {
        let start = Instant::now();

        // Build initial vocabulary
        let bytes: Vec<u8> = training_data.bytes().collect();
        let unique_bytes: std::collections::HashSet<u8> = bytes.iter().cloned().collect();

        for &b in &unique_bytes {
            self.vocab.insert(b as i32, VocabEntry::Byte(b as i32));
        }

        // Convert to byte list
        let mut byte_list: Vec<i32> = bytes.iter().map(|&b| b as i32).collect();
        let mut unused_code_pt = CODE_POINT_START;

        // Byte pair loop
        loop {
            // Count pairs
            let mut pairs: HashMap<Pair, i32> = HashMap::new();
            for i in 0..byte_list.len().saturating_sub(1) {
                let pair = Pair {
                    b1: byte_list[i],
                    b2: byte_list[i + 1],
                };
                *pairs.entry(pair).or_insert(0) += 1;
            }

            if pairs.is_empty() {
                break;
            }

            // Find max frequency pair
            let max_pair = pairs
                .iter()
                .max_by_key(|(_, &count)| count)
                .map(|(pair, _)| *pair)
                .unwrap();

            let max_freq = pairs[&max_pair];

            if max_freq == 1 {
                break;
            }

            // Build new byte list (THIS HAS THE BUG - matches Python exactly)
            // Iterates through pairs, only appends first element or merged token
            // Last element is never appended!
            let mut new_byte_list = Vec::new();
            for i in 0..byte_list.len().saturating_sub(1) {
                if byte_list[i] == max_pair.b1 && byte_list[i + 1] == max_pair.b2 {
                    new_byte_list.push(unused_code_pt);
                } else {
                    new_byte_list.push(byte_list[i]);
                }
            }

            byte_list = new_byte_list;

            // Add to vocabulary
            self.vocab.insert(unused_code_pt, VocabEntry::Pair(max_pair));
            unused_code_pt += 1;
        }

        let elapsed = start.elapsed();
        println!("train took {:.4} seconds", elapsed.as_secs_f64());
    }

    fn encode(&self, s: &str) -> Vec<i32> {
        let start = Instant::now();

        let mut byte_list: Vec<i32> = s.bytes().map(|b| b as i32).collect();
        let mut pair_found = true;

        while pair_found {
            pair_found = false;

            for i in 0..byte_list.len().saturating_sub(1) {
                let s_pair = Pair {
                    b1: byte_list[i],
                    b2: byte_list[i + 1],
                };

                for (&code_pt, vocab_entry) in &self.vocab {
                    if let VocabEntry::Pair(vocab_pair) = vocab_entry {
                        if s_pair == *vocab_pair {
                            byte_list[i] = -1;
                            byte_list[i + 1] = code_pt;
                            pair_found = true;
                            break;
                        }
                    }
                }
            }

            byte_list.retain(|&b| b != -1);
        }

        let elapsed = start.elapsed();
        println!("encode took {:.4} seconds", elapsed.as_secs_f64());

        byte_list
    }

    fn decode(&self, tokens: &[i32]) -> String {
        let start = Instant::now();

        let mut final_code_points: Vec<i32> = tokens.to_vec();
        let mut pair_inserted = true;

        while pair_inserted {
            pair_inserted = false;
            let mut new_list = Vec::new();

            for &pt in &final_code_points {
                if pt <= CODE_POINT_START - 1 {
                    new_list.push(pt);
                } else if let Some(VocabEntry::Pair(pair)) = self.vocab.get(&pt) {
                    new_list.push(pair.b1);
                    new_list.push(pair.b2);
                    pair_inserted = true;
                } else {
                    new_list.push(pt);
                }
            }

            final_code_points = new_list;
        }

        let elapsed = start.elapsed();
        println!("decode took {:.4} seconds", elapsed.as_secs_f64());

        // Convert to bytes then string
        let bytes: Vec<u8> = final_code_points.iter().map(|&b| b as u8).collect();
        String::from_utf8(bytes).unwrap_or_else(|_| String::from("ERROR: Invalid UTF-8"))
    }
}

fn main() {
    println!("\n======================================================================");
    println!("TOKENIZER MAIN EXECUTION (Rust)");
    println!("======================================================================");

    // Read file
    let data = fs::read_to_string("training_text.txt")
        .expect("Error: Could not read training_text.txt");

    let byte_count = data.as_bytes().len();
    println!("INPUT: Loaded from training_text.txt");
    println!("       Length: {} characters, {} bytes", data.len(), byte_count);

    println!("\n--- TRAINING ---");
    let mut tokenizer = Tokenizer::new();
    tokenizer.train(&data);
    println!("Vocabulary size: {} tokens", tokenizer.vocab.len());
    let learned_merges = tokenizer
        .vocab
        .iter()
        .filter(|(&k, _)| k >= CODE_POINT_START)
        .count();
    println!("Learned merges: {}", learned_merges);

    println!("\n--- ENCODING ---");
    let encoded_data = tokenizer.encode(&data);
    println!("Token count: {}", encoded_data.len());
    let compression = (1.0 - encoded_data.len() as f64 / byte_count as f64) * 100.0;
    println!(
        "Compression: {:.1}% ({} bytes → {} tokens)",
        compression,
        byte_count,
        encoded_data.len()
    );

    println!("\n--- DECODING ---");
    let decoded_data = tokenizer.decode(&encoded_data);
    println!("Output length: {} characters", decoded_data.len());

    println!("\n--- RESULT ---");
    if data == decoded_data {
        println!("✓ SUCCESS: Input matches output");
    } else {
        println!("✗ FAILED: Input does not match output");
        println!("  Expected length: {}", data.len());
        println!("  Got length: {}", decoded_data.len());
    }

    println!("======================================================================\n");
}
